File: packages/sdk/clawdstrike-hunt/src/playbook.ts
import { hunt } from './local.js';
import type { HuntOptions } from './local.js';
import { correlate } from './correlate/index.js';
import { IocDatabase } from './correlate/index.js';
import { buildReport, signReport, collectEvidence } from './report.js';
import { parseHumanDuration } from './duration.js';
import type { CorrelationRule, TimelineEvent, Alert, IocMatch, HuntReport, NormalizedVerdict, EvidenceItem } from './types.js';

export interface PlaybookResult {
  events: TimelineEvent[];
  alerts: Alert[];
  iocMatches: IocMatch[];
  report?: HuntReport;
}

interface PlaybookInit {
  start?: string | Date;
  verdictFilter?: NormalizedVerdict;
  rules: CorrelationRule[];
  iocDb?: IocDatabase;
  deduplicateWindow?: number;
  reportTitle?: string;
  signKey?: string;
  huntOptions: Partial<HuntOptions>;
}

export class Playbook {
  private readonly _start?: string | Date;
  private readonly _verdictFilter?: NormalizedVerdict;
  private readonly _rules: CorrelationRule[];
  private readonly _iocDb?: IocDatabase;
  private readonly _deduplicateWindow?: number;
  private readonly _reportTitle?: string;
  private readonly _signKey?: string;
  private readonly _huntOptions: Partial<HuntOptions>;

  private constructor(init: Partial<PlaybookInit> = {}) {
    this._start = init.start;
    this._verdictFilter = init.verdictFilter;
    this._rules = init.rules ?? [];
    this._iocDb = init.iocDb;
    this._deduplicateWindow = init.deduplicateWindow;
    this._reportTitle = init.reportTitle;
    this._signKey = init.signKey;
    this._huntOptions = init.huntOptions ?? {};
  }

  static create(): Playbook {
    return new Playbook();
  }

  since(timeRange: string | Date): Playbook {
    return new Playbook({ ...this.toInit(), start: timeRange });
  }

  filter(verdict: NormalizedVerdict): Playbook {
    return new Playbook({ ...this.toInit(), verdictFilter: verdict });
  }

  correlate(rules: CorrelationRule[]): Playbook {
    return new Playbook({ ...this.toInit(), rules });
  }

  enrich(iocDb: IocDatabase): Playbook {
    return new Playbook({ ...this.toInit(), iocDb });
  }

  deduplicate(window: string | number): Playbook {
    let windowMs: number;
    if (typeof window === 'string') {
      windowMs = parseHumanDuration(window) ?? 0;
    } else {
      windowMs = window;
    }
    return new Playbook({ ...this.toInit(), deduplicateWindow: windowMs });
  }

  report(title: string): Playbook {
    return new Playbook({ ...this.toInit(), reportTitle: title });
  }

  sign(keyHex: string): Playbook {
    return new Playbook({ ...this.toInit(), signKey: keyHex });
  }

  async run(): Promise<PlaybookResult> {
    // 1. Hunt events
    let events = await hunt({ ...this._huntOptions, start: this._start });

    // 2. Filter by verdict
    if (this._verdictFilter) {
      events = events.filter(e => e.verdict === this._verdictFilter);
    }

    // 3. Correlate
    let alerts: Alert[] = [];
    if (this._rules.length > 0) {
      alerts = correlate(this._rules, events);
    }

    // 4. IOC enrichment
    let iocMatches: IocMatch[] = [];
    if (this._iocDb) {
      iocMatches = this._iocDb.matchEvents(events);
    }

    // 5. Deduplicate alerts
    if (this._deduplicateWindow !== undefined && this._deduplicateWindow > 0) {
      alerts = deduplicateAlerts(alerts, this._deduplicateWindow);
    }

    // 6. Build report
    let report: HuntReport | undefined;
    if (this._reportTitle) {
      const evidenceSources: (Alert | TimelineEvent[] | IocMatch[])[] = [];
      for (const a of alerts) {
        evidenceSources.push(a);
      }
      if (events.length > 0) {
        evidenceSources.push(events);
      }
      if (iocMatches.length > 0) {
        evidenceSources.push(iocMatches);
      }
      if (evidenceSources.length > 0) {
        const evidence = deduplicateEvidence(collectEvidence(...evidenceSources));
        if (evidence.length > 0) {
          report = buildReport(this._reportTitle, evidence);
          // 7. Sign
          if (this._signKey) {
            report = await signReport(report, this._signKey);
          }
        }
      }
    }

    return { events, alerts, iocMatches, report };
  }

  toJSON(): Record<string, unknown> {
    return {
      start: this._start instanceof Date ? this._start.toISOString() : this._start,
      verdictFilter: this._verdictFilter,
      rules: this._rules,
      deduplicateWindow: this._deduplicateWindow,
      reportTitle: this._reportTitle,
      huntOptions: this._huntOptions,
    };
  }

  static fromJSON(json: Record<string, unknown>): Playbook {
    return new Playbook({
      start: json.start as string | undefined,
      verdictFilter: json.verdictFilter as NormalizedVerdict | undefined,
      rules: json.rules as CorrelationRule[] | undefined,
      deduplicateWindow: json.deduplicateWindow as number | undefined,
      reportTitle: json.reportTitle as string | undefined,
      huntOptions: json.huntOptions as Partial<HuntOptions> | undefined,
    });
  }

  private toInit(): PlaybookInit {
    return {
      start: this._start,
      verdictFilter: this._verdictFilter,
      rules: this._rules,
      iocDb: this._iocDb,
      deduplicateWindow: this._deduplicateWindow,
      reportTitle: this._reportTitle,
      signKey: this._signKey,
      huntOptions: this._huntOptions,
    };
  }
}

function deduplicateEvidence(items: EvidenceItem[]): EvidenceItem[] {
  const seen = new Set<string>();
  const result: EvidenceItem[] = [];
  let nextIndex = 0;
  for (const item of items) {
    const key = `${item.sourceType}|${item.timestamp instanceof Date ? item.timestamp.toISOString() : String(item.timestamp)}|${item.summary}`;
    if (!seen.has(key)) {
      seen.add(key);
      result.push({ ...item, index: nextIndex });
      nextIndex++;
    }
  }
  return result;
}

function deduplicateAlerts(alerts: Alert[], windowMs: number): Alert[] {
  const seen = new Map<string, Date>();
  return alerts.filter(alert => {
    const key = alert.ruleName;
    const lastSeen = seen.get(key);
    if (lastSeen && (alert.triggeredAt.getTime() - lastSeen.getTime()) < windowMs) {
      return false;
    }
    seen.set(key, alert.triggeredAt);
    return true;
  });
}

File: packages/sdk/hush-py/src/clawdstrike/hunt/playbook.py
"""Hunt Playbook — composable builder for hunt workflows.

Provides an immutable builder that chains hunt, correlate, enrich,
deduplicate, and report generation steps.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timedelta
from typing import Any

from clawdstrike.hunt.correlate import correlate
from clawdstrike.hunt.duration import parse_human_duration
from clawdstrike.hunt.ioc import IocDatabase
from clawdstrike.hunt.local import hunt
from clawdstrike.hunt.report import build_report, collect_evidence, sign_report
from clawdstrike.hunt.types import (
    Alert,
    CorrelationRule,
    HuntReport,
    IocMatch,
    NormalizedVerdict,
    TimelineEvent,
)


@dataclass(frozen=True)
class PlaybookResult:
    """Result of a playbook execution."""

    events: tuple[TimelineEvent, ...]
    alerts: tuple[Alert, ...]
    ioc_matches: tuple[IocMatch, ...]
    report: HuntReport | None = None


def _deduplicate_alerts(alerts: list[Alert], window: timedelta) -> list[Alert]:
    """Remove alerts with the same rule_name within the dedup window."""
    seen: dict[str, datetime] = {}
    result: list[Alert] = []
    for alert in alerts:
        key = alert.rule_name
        last_seen = seen.get(key)
        if last_seen is not None and (alert.triggered_at - last_seen) < window:
            continue
        seen[key] = alert.triggered_at
        result.append(alert)
    return result


class Playbook:
    """Immutable builder for composing hunt workflows."""

    def __init__(
        self,
        *,
        start: datetime | str | None = None,
        verdict_filter: NormalizedVerdict | None = None,
        rules: tuple[CorrelationRule, ...] = (),
        ioc_db: IocDatabase | None = None,
        deduplicate_window: timedelta | None = None,
        report_title: str | None = None,
        sign_key: str | None = None,
        hunt_kwargs: dict[str, Any] | None = None,
    ) -> None:
        self._start = start
        self._verdict_filter = verdict_filter
        self._rules = rules
        self._ioc_db = ioc_db
        self._deduplicate_window = deduplicate_window
        self._report_title = report_title
        self._sign_key = sign_key
        self._hunt_kwargs: dict[str, Any] = hunt_kwargs or {}

    def _copy(self, **overrides: Any) -> Playbook:
        base = {
            "start": self._start,
            "verdict_filter": self._verdict_filter,
            "rules": self._rules,
            "ioc_db": self._ioc_db,
            "deduplicate_window": self._deduplicate_window,
            "report_title": self._report_title,
            "sign_key": self._sign_key,
            "hunt_kwargs": self._hunt_kwargs,
        }
        base.update(overrides)
        return Playbook(**base)

    @classmethod
    def create(cls) -> Playbook:
        return cls()

    def since(self, time_range: datetime | str) -> Playbook:
        return self._copy(start=time_range)

    def filter(self, verdict: NormalizedVerdict) -> Playbook:
        return self._copy(verdict_filter=verdict)

    def correlate(self, rules: list[CorrelationRule]) -> Playbook:
        return self._copy(rules=tuple(rules))

    def enrich(self, ioc_db: IocDatabase) -> Playbook:
        return self._copy(ioc_db=ioc_db)

    def deduplicate(self, window: timedelta | str) -> Playbook:
        if isinstance(window, str):
            parsed = parse_human_duration(window)
            if parsed is None:
                parsed = timedelta(0)
            return self._copy(deduplicate_window=parsed)
        return self._copy(deduplicate_window=window)

    def report(self, title: str) -> Playbook:
        return self._copy(report_title=title)

    def sign(self, key_hex: str) -> Playbook:
        return self._copy(sign_key=key_hex)

    def run(self) -> PlaybookResult:
        """Execute the playbook pipeline and return results."""
        # 1. Hunt events
        events = hunt(start=self._start, **self._hunt_kwargs)

        # 2. Filter by verdict
        if self._verdict_filter is not None:
            events = [e for e in events if e.verdict == self._verdict_filter]

        # 3. Correlate
        alerts: list[Alert] = []
        if self._rules:
            alerts = correlate(list(self._rules), events)

        # 4. IOC enrichment
        ioc_matches: list[IocMatch] = []
        if self._ioc_db is not None:
            ioc_matches = self._ioc_db.match_events(events)

        # 5. Deduplicate
        if self._deduplicate_window is not None and self._deduplicate_window > timedelta(0):
            alerts = _deduplicate_alerts(alerts, self._deduplicate_window)

        # 6. Build report
        hunt_report: HuntReport | None = None
        if self._report_title is not None:
            evidence = collect_evidence(*alerts, events, ioc_matches)
            if evidence:
                hunt_report = build_report(self._report_title, evidence)
                # 7. Sign
                if self._sign_key is not None and hunt_report is not None:
                    hunt_report = sign_report(hunt_report, self._sign_key)

        return PlaybookResult(
            events=tuple(events),
            alerts=tuple(alerts),
            ioc_matches=tuple(ioc_matches),
            report=hunt_report,
        )

    def to_dict(self) -> dict[str, Any]:
        start_val = self._start
        if isinstance(start_val, datetime):
            start_val = start_val.isoformat()
        return {
            "start": start_val,
            "verdict_filter": self._verdict_filter.value if self._verdict_filter else None,
            "rules": [r.name for r in self._rules],
            "deduplicate_window": self._deduplicate_window.total_seconds() if self._deduplicate_window else None,
            "report_title": self._report_title,
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> Playbook:
        return cls(
            start=data.get("start"),
            verdict_filter=NormalizedVerdict(data["verdict_filter"]) if data.get("verdict_filter") else None,
            deduplicate_window=timedelta(seconds=data["deduplicate_window"]) if data.get("deduplicate_window") else None,
            report_title=data.get("report_title"),
        )


__all__ = ["Playbook", "PlaybookResult"]
